### PR TITLE
feat(Player): add setNode & replayOnDc

### PR DESCRIPTION
**Pull request by KagChi**
setNode is used to move the player to another connected node

replayOnDc is used when the node disconnect, music will be replayed on another connected node

### CODE REVIEW

File: src/structures/Player.ts [hunk 1]
```diff
@@ -139,15 +146,37 @@ export class Player {
   ): Promise<SearchResult> {
     return this.manager.search(query, requester);
   }
+  /**
+   * Move the player to another connected node
+   * @param name
+   */
+  async movePlayer(name: string) {
+    const node = this.manager.nodes.get(name)
+    if(!node.connected) throw Error('The node is not connected');
```
Comment (Solaris9) [inline src/structures/Player.ts:155]: If node does not exist this will throw an error.

File: src/structures/Player.ts [hunk 2]
```diff
@@ -139,15 +146,37 @@ export class Player {
   ): Promise<SearchResult> {
     return this.manager.search(query, requester);
   }
+  /**
+   * Move the player to another connected node
+   * @param name
+   */
+  async movePlayer(name: string) {
```
Comment (Solaris9) [inline src/structures/Player.ts:153]: `setNode` would be better naming.

File: src/structures/Manager.ts [hunk 3]
```diff
@@ -491,6 +491,8 @@ export interface ManagerOptions {
   clientId?: string;
   /** Value to use for the `Client-Name` header. */
   clientName?: string;
+  /** Auto replay whenever node disconnect */
+  ReplayOnDc?: boolean;
```
Comment (Solaris9) [inline src/structures/Manager.ts:495]: Should be camcalCase instead of PascalCase.

File: src/structures/Player.ts [hunk 4]
```diff
@@ -139,15 +146,37 @@ export class Player {
...
+    const node = this.manager.nodes.get(name)
+    if(!node.connected) throw Error('The node is not connected');
+    if (!name || node.options.identifier === this.node.options.identifier) return this;
```
Comment (Solaris9) [inline src/structures/Player.ts:156]: The left hand side condition should be before accessing the Node members and the name parameter would be the [identifier](https://github.com/MenuDocs/erela.js/blob/master/src/structures/Node.ts#L118), simply:
```javascript
if (this.node.options.identifier === name) {}
```

File: src/structures/Player.ts [hunk 5]
```diff
@@ -139,15 +146,37 @@ export class Player {
...
+    if(!node.connected) throw Error('The node is not connected');
+    if (!name || node.options.identifier === this.node.options.identifier) return this;  
+    const options = {
+      op: "play",
+      guildId: this.guild,
+      track: this.queue.current.track,
+      startTime: this.position,
+    };
 
+    if (typeof options.track !== "string") {
```
Comment (Solaris9) [inline src/structures/Player.ts:164]: This will always be a string as Track#track is a string.

File: src/structures/Player.ts [hunk 6]
```diff
@@ -139,15 +146,37 @@ export class Player {
...
+      options.track = (options.track as Track).track;
```
Comment (Solaris9) [inline src/structures/Player.ts:165]: As the above comment, `options.track` cannot be anything other than a string.

File: src/structures/Player.ts [hunk 7]
```diff
@@ -139,15 +146,37 @@ export class Player {
...
+    }
+    this.node = node;
+    this.manager.players.set(this.guild, this)
```
Comment (Solaris9) [inline src/structures/Player.ts:168]: This is unneeded as the player is already in the players map.

File: src/structures/Player.ts [hunk 8]
```diff
@@ -139,15 +146,37 @@ export class Player {
...
+    await this.node.send({ op: "voiceUpdate", guildId: this.voiceState.guildId, sessionId: this.voiceState.sessionId, event: { token: this.voiceState.event.token, endpoint: this.voiceState.event.endpoint } })
```
Comment (Solaris9) [inline src/structures/Player.ts:169]: You can use the variable `Player#voiceState` ([type](https://github.com/MenuDocs/erela.js/blob/master/src/structures/Utils.ts#L322), [usage](https://github.com/MenuDocs/erela.js/blob/master/src/structures/Manager.ts#L457-L463)) instead as it has all the info required to minimize the use of named properties

**Review by Solaris9 — CHANGES_REQUESTED** [review]
I'm not sure how to review an entire file but the `docs.json` is not for typedoc but rather a documentation generator I was working on while I was still the project owner, this file holds the guide's info.

File: src/structures/Player.ts [hunk 9]
```diff
@@ -126,6 +126,13 @@ export class Player {
     this.manager.players.set(options.guild, this);
     this.manager.emit("playerCreate", this);
     this.setVolume(options.volume ?? 100);
+    this.manager.on('nodeDisconnect', (node) => {
+      if(this.manager.options.ReplayOnDc && this.manager.nodes.size > 1) {
+        for(const players of [...this.manager.players.filter(x => x.node === node).values()]) {
+          players.movePlayer(this.manager.leastUsedNodes.first().options.identifier)
```
Comment (Solaris9) [inline src/structures/Player.ts:132]: This reference still has to be renamed.

**Review by Solaris9 — CHANGES_REQUESTED** [review]

File: src/structures/Player.ts [hunk 10]
```diff
@@ -139,15 +146,33 @@ export class Player {
   ): Promise<SearchResult> {
     return this.manager.search(query, requester);
   }
-
+  /**
+   * Move the player to another connected node
+   * @param name
+   */
+  async setNode(name: string) {
```
Comment (Anish-Shobith) [inline src/structures/Player.ts:153]: Please add a return type

File: src/structures/Player.ts (hunk 1 above)
Comment (Anish-Shobith) [reply src/structures/Player.ts:156]: i would say check if there is more than one node and check if they are connected then filter the least used nodes and connect to it any play, if there is then only you are able to change the node.

File: src/structures/Player.ts [hunk 11]
```diff
@@ -126,6 +126,13 @@ export class Player {
     this.manager.players.set(options.guild, this);
     this.manager.emit("playerCreate", this);
     this.setVolume(options.volume ?? 100);
+    this.manager.on('nodeDisconnect', (node) => {
```
Comment (MeLike2D) [inline src/structures/Player.ts:129]: Possibly have an if statement wrapping this? If `replayOnDc` is false you'll just be adding a redundant listener

**Comment by txj-xyz** [comment]
Any updates on this? Would like to see this implemented some time soon.

**Pull request closed without merging**